crates/misc/component-fuzz-util/src/lib.rs

```diff
@@ -553,112 +553,135 @@ pub fn rust_type(ty: &Type, name_counter: &mut u32, declarations: &mut TokenStre
     }
 }
 
-fn make_component_name(name_counter: &mut u32) -> String {
-    let name = format!("$Foo{name_counter}");
-    *name_counter += 1;
-    name
+struct TypesBuilder<'a> {
+    next: u32,
+    worklist: Vec<(u32, &'a Type)>,
 }
 
-fn write_component_type(
-    ty: &Type,
-    f: &mut String,
-    name_counter: &mut u32,
-    declarations: &mut String,
-) {
-    match ty {
-        Type::Unit => f.push_str("unit"),
-        Type::Bool => f.push_str("bool"),
-        Type::S8 => f.push_str("s8"),
-        Type::U8 => f.push_str("u8"),
-        Type::S16 => f.push_str("s16"),
-        Type::U16 => f.push_str("u16"),
-        Type::S32 => f.push_str("s32"),
-        Type::U32 => f.push_str("u32"),
-        Type::S64 => f.push_str("s64"),
-        Type::U64 => f.push_str("u64"),
-        Type::Float32 => f.push_str("float32"),
-        Type::Float64 => f.push_str("float64"),
-        Type::Char => f.push_str("char"),
-        Type::String => f.push_str("string"),
-        Type::List(ty) => {
-            let mut case = String::new();
-            write_component_type(ty, &mut case, name_counter, declarations);
-            let name = make_component_name(name_counter);
-            write!(declarations, "(type {name} (list {case}))").unwrap();
-            f.push_str(&name);
-        }
-        Type::Record(types) => {
-            let mut fields = String::new();
-            for (index, ty) in types.iter().enumerate() {
-                write!(fields, r#" (field "f{index}" "#).unwrap();
-                write_component_type(ty, &mut fields, name_counter, declarations);
-                fields.push_str(")");
+impl<'a> TypesBuilder<'a> {
+    fn write_ref(&mut self, ty: &'a Type, dst: &mut String) {
+        match ty {
+            // Primitive types can be referenced directly
+            Type::Unit => dst.push_str("unit"),
+            Type::Bool => dst.push_str("bool"),
+            Type::S8 => dst.push_str("s8"),
+            Type::U8 => dst.push_str("u8"),
+            Type::S16 => dst.push_str("s16"),
+            Type::U16 => dst.push_str("u16"),
+            Type::S32 => dst.push_str("s32"),
+            Type::U32 => dst.push_str("u32"),
+            Type::S64 => dst.push_str("s64"),
+            Type::U64 => dst.push_str("u64"),
+            Type::Float32 => dst.push_str("float32"),
+            Type::Float64 => dst.push_str("float64"),
+            Type::Char => dst.push_str("char"),
+            Type::String => dst.push_str("string"),
+
+            // Otherwise the items is deferred to get a type declaration to
+            // itself and it's referred to by index.
```

Comment:
I can't read this comment. How about "Otherwise emit a reference to the type and remember to generate the corresponding type alias later"?